### ISSUE TITLE
Global callbacks

### ISSUE THREAD

**Issue by fabn** [aasm/aasm#253]
It would be nice (mostly for auditing purpose) to have a set of global callbacks in this gem, something like

```ruby
class Job
  include AASM
  aasm before_state_change: :log_it, before_event: :log_it_too do
    state :sleeping, :initial => true
    state :running
    event :run do
      transitions :from => :sleeping, :to => :running
    end
  end

  def log_it
    logger.info "triggered #{aasm.current_event}"
    logger.info "from #{aasm.from_state} to #{aasm.to_state}"
  end
end
```

What do you think?

**Comment by jeremywadsack** [comment]
+1 I was just looking at how to do this so that we can track even transition timing (and other such metrics).

**Comment by jeremywadsack** [comment]
This looks similar to #221 and #108 which is what I really want. Because I want to decorate an existing AASM model (defined in a shared gem) with callbacks (that are specific to the app) but apply the same code all transitions.

**Comment by rposborne** [comment]
+1

**Comment by alto** [comment]
Okay, I see your point. Will work on that from now on.

**Comment by alto** [comment]
Would this be sufficient?

```ruby
  aasm do
    after_all_transitions :do_logging_or_whatever
  end

  def do_logging_or_whatever
    ...
  end
```

In `do_logging_or_whatever` you can access `aasm.from_state`, `aasm.to_state` and `aasm.current_event`.

**Comment by jeremywadsack** [comment]
Yeah, that works for me.

**Comment by fabn** [comment]
That would be good. Imho also a before callback would be nice to have a trace in case something goes wrong during transition.

**Comment by anilmaurya** [comment]
@alto are you working on this feature ? If not then I can help you on this.

**Comment by alto** [comment]
@anilmaurya Thanks for your offer. I'm sorry, I'm already working on it.

**Comment by anilmaurya** [comment]
@alto No problem . Happy coding :smile:

**Comment by alto** [comment]
Please take a look at #270 and let me know if that works for your problem. Will release soon.

**Comment by fabn** [comment]
Awesome job, thanks.

**Comment by jeremywadsack** [comment]
Yeah, that works for me too. Thanks.

**Comment by alto** [comment]
Closed with #270 .

**Comment by alto** [comment]
I just released version `4.4.0` including the change.
